pkg/alertmanager/amcfg.go

```diff
@@ -1226,21 +1246,30 @@ func (hc *httpClientConfig) sanitize(amVersion semver.Version, logger log.Logger
 	if hc == nil {
 		return
 	}
-	// we don't need to do any sanitization in this case and return early
-	if hc.Authorization == nil {
-		return
-	}
 
-	if hc.BasicAuth != nil {
-		msg := "'basicAuth' and 'authorization' are mutually exclusive, 'basicAuth' has taken precedence"
-		level.Warn(logger).Log("msg", msg)
+	if hc.Authorization != nil && !amVersion.GTE(semver.MustParse("0.22.0")) {
+		msg := "'authorization' set in 'http_config' but  supported in AlertManager >= 0.22.0 only - dropping field from provided config"
+		level.Warn(logger).Log("msg", msg, "current_version", amVersion.String())
 		hc.Authorization = nil
 	}
-	// we could have returned here but useful to grab the log and bubble up the warning
-	if httpAuthzAllowed := amVersion.GTE(semver.MustParse("0.22.0")); !httpAuthzAllowed {
-		msg := "'authorization' set in 'http_config' but  supported in AlertManager >= 0.22.0 only - dropping field from provided config"
+
+	if hc.OAuth2 != nil && !amVersion.GTE(semver.MustParse("0.22.0")) {
+		msg := "'oauth2' set in 'http_config' but  supported in AlertManager >= 0.22.0 only - dropping field from provided config"
+		level.Warn(logger).Log("msg", msg, "current_version", amVersion.String())
+		hc.OAuth2 = nil
+	}
+
+	if hc.FollowRedirects != nil && !amVersion.GTE(semver.MustParse("0.22.0")) {
+		msg := "'follow_redirects' set in 'http_config' but  supported in AlertManager >= 0.22.0 only - dropping field from provided config"
 		level.Warn(logger).Log("msg", msg, "current_version", amVersion.String())
+		hc.FollowRedirects = nil
+	}
+
+	if hc.BasicAuth != nil && (hc.OAuth2 != nil || hc.Authorization != nil) {
```

Comment:
don't we need to check for `hc.OAuth2` and `hc.Authorization` being not nil simultaneously? Also wondering if we shouldn't return an error instead since it's not like you want to use a field that isn't supported.

Reply:
In regards to the error, I was returning an error here originally but I believe we discussed that offline and reverted since right now in https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/alertmanager/amcfg.go#L1243 we are simply dropping the field and we were aligning the behaviour.

Saying that, my initial instinct here was that the right thing to do was to error so maybe we should do a u-turn? :)

Reply:
sorry for introducing confusion then :) I agree that returning an error makes more sense since the operator can't really decide which option makes sense than the others.